File: native/android/app/src/main/kotlin/art/lazying/auspice/VisionCapture.kt
package art.lazying.auspice

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*

data class LandmarkFrame(val points: List<List<Double>>, val width: Int, val height: Int, val timestamp: Long) {
    fun json() = buildJsonObject {
        put("width",width); put("height",height); put("timestamp",timestamp)
        put("landmarks",JsonArray(points.map { p -> buildJsonObject { put("x",p[0]);put("y",p[1]);put("z",p[2]) } }))
    }
}

/** Bounded, transient burst. All geometry rules are in the shared engine. */
class CaptureWindow {
    private val frames = mutableListOf<LandmarkFrame>()
    private var receivedAt = 0L
    val ready get() = frames.size >= 8 && frames.last().timestamp-frames.first().timestamp >= 650
    fun clear() { frames.clear(); receivedAt=0 }
    fun append(frame: LandmarkFrame, now: Long = System.nanoTime()/1_000_000) {
        if(frame.points.isEmpty() || frame.width<=0 || frame.height<=0) { clear();return }
        frames.lastOrNull()?.let { last ->
            if(frame.timestamp<=last.timestamp || frame.timestamp-last.timestamp>300 || frame.width!=last.width || frame.height!=last.height) clear()
        }
        frames.add(frame)
        while(frames.size>12 || frame.timestamp-frames.first().timestamp>1800) frames.removeAt(0)
        receivedAt=now
    }
    fun snapshot(now: Long = System.nanoTime()/1_000_000): JsonArray =
        if(ready && now-receivedAt<=500) JsonArray(frames.map { it.json() }) else JsonArray(emptyList())
}

@Serializable
data class VisionMeasurement(val version: Int=2,val samples: Int=8,val typeCandidates: List<String> = emptyList(),val limitations: List<String> = emptyList())

fun visionError(error: Throwable): String = error.message?.takeIf { it.startsWith("vision.") }?.let { t(it) }
    ?: l("This reading could not be computed. Please try again.")

fun visionState(value: String): String = if(value=="uncertain") t("vision.uncertain") else l(value)

File: native/ios/Auspice/VisionCapture.swift
import Foundation

/// Transient landmarks only: no pictures, matching identifiers or previous
/// person's measurements. Geometry/stability rules live in the shared engine.
struct LandmarkFrame {
    var landmarks: [[String: Double]]
    var width: Int
    var height: Int
    var timestamp: Int
    var dictionary: [String: Any] {
        ["landmarks": landmarks, "width": width, "height": height, "timestamp": timestamp]
    }
}

struct CaptureWindow {
    private(set) var frames: [LandmarkFrame] = []
    private var receivedAt = 0.0
    var ready: Bool { frames.count >= 8 && (frames.last!.timestamp - frames.first!.timestamp) >= 650 }
    mutating func clear() { frames = []; receivedAt = 0 }
    mutating func append(_ frame: LandmarkFrame, now: Double = ProcessInfo.processInfo.systemUptime) {
        guard !frame.landmarks.isEmpty, frame.width > 0, frame.height > 0 else { clear(); return }
        if let last = frames.last, frame.timestamp <= last.timestamp || frame.timestamp - last.timestamp > 300 ||
            frame.width != last.width || frame.height != last.height { clear() }
        frames.append(frame)
        frames = Array(frames.filter { frame.timestamp - $0.timestamp <= 1800 }.suffix(12))
        receivedAt = now
    }
    func snapshot(now: Double = ProcessInfo.processInfo.systemUptime) -> [[String: Any]] {
        guard ready, now - receivedAt <= 0.5 else { return [] }
        return frames.map(\.dictionary)
    }
}

struct VisionMeasurement: Codable {
    var version: Int
    var samples: Int
    var typeCandidates: [String]
    var limitations: [String]
}

func visionError(_ error: Error) -> String {
    let key = error.localizedDescription
    return key.hasPrefix("vision.") ? t(key) : l("This reading could not be computed. Please try again.")
}

func visionState(_ value: String) -> String { value == "uncertain" ? t("vision.uncertain") : l(value) }
